ports/espressif/common-hal/analogbufio/BufferedIn.c

```diff
@@ -0,0 +1,283 @@
+    if(output_format == ADC_DIGI_OUTPUT_FORMAT_TYPE2) {
+        if (data->type2.channel >= SOC_ADC_CHANNEL_NUM(unit)) return false;
+        if (adc_channel != data->type2.channel) return false;
+    } else {
+        if( convert_mode == ADC_CONV_SINGLE_UNIT_1 ) {
+            unit = 0;
+        } else {
+            unit = 1;
+        }
+        if (data->type1.channel >= SOC_ADC_CHANNEL_NUM(unit)) return false;
+        if (adc_channel != data->type1.channel) return false;
+    }
+    if (unit > 2) return false;
+    return true;
+}
+
+
+uint32_t common_hal_analogbufio_bufferedin_readinto(analogbufio_bufferedin_obj_t *self, uint8_t *buffer, uint32_t len, uint8_t bytes_per_sample) {
+    uint8_t result[NUM_SAMPLES_PER_INTERRUPT] __attribute__ ((aligned (4))) = {0};
+    uint32_t captured_samples = 0;
+    uint32_t captured_bytes = 0;
+    esp_err_t ret;
+    uint32_t ret_num = 0;
+
+#if defined(DEBUG_ANALOGBUFIO)
+    mp_printf(&mp_plat_print,"Required bytes: %d\n",len);
+#endif //DEBUG_ANALOGBUFIO
+
+    while(captured_bytes < len) {
+        ret_num = 0;
+        ret = adc_digi_read_bytes(result, NUM_SAMPLES_PER_INTERRUPT, &ret_num, ADC_READ_TIMEOUT_MS);
+
+        if (ret == ESP_OK) {
+            for(uint32_t i=0; i<ret_num; i+=ADC_RESULT_BYTE) {
+                if(check_valid_data((adc_digi_output_data_t *) (void *)&result[i])) {
+                    if(captured_bytes < len) {
+                        if(ADC_RESULT_BYTE == 2) {
+                            if(output_format == ADC_DIGI_OUTPUT_FORMAT_TYPE1) {
+                                *(uint16_t *)(void *)&buffer[captured_bytes] = ((adc_digi_output_data_t *) (void *)&result[i])->type1.data;
+                            } else {
+                                *(uint16_t *)(void *)&buffer[captured_bytes] = ((adc_digi_output_data_t *) (void *)&result[i])->type2.data;
+                            }
+                        } else {
+                            if(output_format == ADC_DIGI_OUTPUT_FORMAT_TYPE1) {
+                                *(uint32_t *)(void *)&buffer[captured_bytes] = ((adc_digi_output_data_t *) (void *)&result[i])->type1.data;
+                            } else {
+                                *(uint32_t *)(void *)&buffer[captured_bytes] = ((adc_digi_output_data_t *) (void *)&result[i])->type2.data;
+                            }
+                        }
+                        captured_bytes += ADC_RESULT_BYTE;
+                        captured_samples++;
+                    } else {
+                        return captured_samples;
+                    }
+                } else {
+#if defined(DEBUG_ANALOGBUFIO)
+                    if(ADC_RESULT_BYTE == 2) {
+                        mp_printf(&mp_plat_print,"Invalid sample received: 0x%0x\n",*(uint16_t *)(void *)&result[i]);
+                    } else {
+                        mp_printf(&mp_plat_print,"Invalid sample received: 0x%0x\n",*(uint32_t *)(void *)&result[i]);
+                    }
```

Comment:
Can you avoid this checking by validating the buffer data width?

Reply:
I have simplified this check in commit https://github.com/adafruit/circuitpython/pull/7602/commits/ead5751803ae2c770ec5923c5c996cb1f33aa930